Ветка для разработки; sandbox-версия стейт-машины

KCherkasov opened the pull request:
-диаграмма состояний: https://www.lucidchart.com/invitations/accept/d1c33eee-9c3c-42e5-9817-7c3f33c16d8e
-часть кода, реализующая машину состояний (сами конкретные состояния пока без бизнес-логики)
-идея использования стейт-машины - на каждое подключение создавать коннекшн (коннекшны хранятся в ассоциативном контейнере с id сессии в роли ключа), хранящий юзердату, id сессии и экземпляр стейт-машины. Входящие пакеты содержат id сессии и маршрутизируются на соотв. коннекшн. Далее пакет пропускается через стейт-машину в поисках стейта(ов), который(е) знает(ют), как на этот пакет реагировать.
-класс Main - мини-тест корректности реализации стейт-машины (интерфейс Packet выполнен в виде стаба, а конкретные классы стейтов по сути пока тоже стабы)

KCherkasov committed:
Readme upd

цитата отсюда - https://park.mail.ru/blog/topic/view/9596/

KCherkasov committed:
develop branch initial commit; State machine sandbox

created development branch

state machine for KVVArtet team project with a sandbox example of usage
(states logic will be added later)

KCherkasov committed:
readme upd (team members added, description edited)

KCherkasov committed:
one more minor fix

KCherkasov committed:
final fix

KCherkasov committed:
readme upd

KCherkasov committed:
minor codestyle fixing; character creation state preset added

-fixed AbstractState.ID->AbstractState.StateId renaming issues;
-added new state preset for character creation

Badya commented on src/main/java/states/AbstractState.java:
```diff
@@ -0,0 +1,14 @@
+
+import packets.Packet;
+
+public interface AbstractState {
```
Обычно, `Abstract` в названии класса явно свидетельствует о том что это абстрактный класс, а не интерфейс - рекомендую поменять имена `State` и `AbstractState` местами

Badya commented on src/main/java/statemachine/StateStack.java:
```diff
@@ -0,0 +1,117 @@
+    public void pushState(AbstractState.StateId stateId) {
+        this.pendingChanges.add(new PendingChange(StackActions.SA_PUSH, stateId));
+        final String stateName = stateId.toString();
+        System.out.println("push request for state " + stateName + " enqueued");
```
Использование `System.out/err` не гуд.
Нужно прикрутить логгирование, в случае вопросов - пишите в чат

KCherkasov committed:
State and AbstractState names swapped

-requested changes #1 implemented

KCherkasov commented:
Имена интерфейса и абстракта поменял, над логгированием с утра поработаю

Badya commented:
Не забудьте подтянуть к себе HEAD java-park-mail-ru:develop или сделать rebase

KCherkasov committed:
Merge https://github.com/KCherkasov/KVVArtet-09-2017 into develop

KCherkasov committed:
slf4j logger implemented

-requested changes #2 done

KCherkasov commented:
Прикрутил логгер вместо стандартных потоков.

Badya commented on src/main/java/states/AbstractState.java:
```diff
@@ -3,48 +3,44 @@
   }
 
   private final PendingStack stack;
+  protected static final Logger logger = LoggerFactory.getLogger(AbstractState.class);
```
Этот логгер будет даже из конечных классов писать как `AbstractState` - будет создавать путаницу, можно заменить на `getClass()`

KCherkasov replied:
А как именно? Может целесообразнее прикрутить логгер не к абстракту, а к конечным классам тогда? (не понимаю, как использовать нестатический <code>getClass()</code> при инициализации логгера, да и логгер по сути один на всех потомков получается)

KSolovyev replied:
Класс используется, как дополнительная метка при логировании. Те он вам напишет [14:59:59] AbstractState.class: blablabla
И будет абсолютно непонятно, откуда это пришло.
getClass() сделать и правда не получится, но не поленитесь, добавьте по логгеру в каждый класс

Badya commented on src/main/java/statemachine/StateStack.java:
```diff
@@ -114,4 +115,5 @@ StackActions getAction() {
 
     final LinkedList<PendingChange> pendingChanges = new LinkedList<>();
```
Вместо `LinkedList<E>` в объявлении уместнее было бы наверное сделать `Deque<E>`

KCherkasov committed:
Merge branch 'develop' into develop

KCherkasov committed:
travis file added

KCherkasov committed:
Merge branch 'develop' of https://github.com/KCherkasov/KVVArtet-09-2017 into develop

KCherkasov committed:
travis file misspelling fix

KCherkasov committed:
fixing travis check issues

KCherkasov committed:
fixing travis check issues

KCherkasov committed:
one more travis check fixup

Badya commented on src/main/java/states/AbstractState.java:
```diff
@@ -3,48 +3,44 @@
       return true;
   }
 
   private final PendingStack stack;
```
Поля, даже приватные, лучше располагать вверху

KSolovyev commented on src/main/java/states/StateFactory.java:
```diff
@@ -0,0 +1,76 @@
+import org.slf4j.LoggerFactory;
+import statemachine.PendingStack;
+
+public final class StateFactory {
```
Идея со стейтами - хорошая, но фабрика должна быть Спринговым сервисом, а не хелпером со статическими методами. Иначе прокидывать зависимости будет очень сложно

Badya replied:
up

KSolovyev commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,18 @@
+public class Main {
+    public static void main(String[] args) {
+        // тут пример-пустышка - создаем стейт-машину и начинаем играть со стейтами
+        final PendingStack stack = new StateStack();
```
это должен быть тест. Можете попробовать сами писать их, unit-тесты пишутся довольно просто

Badya replied:
up

KSolovyev commented:
> -идея использования стейт-машины - на каждое подключение создавать коннекшн (коннекшны хранятся в ассоциативном контейнере с id сессии в роли ключа), хранящий юзердату, id сессии и экземпляр стейт-машины. Входящие пакеты содержат id сессии и маршрутизируются на соотв. коннекшн. Далее пакет пропускается через стейт-машину в поисках стейта(ов), который(е) знает(ют), как на этот пакет реагировать.

Обычно у вас есть некоторая иговая сессия, реализующая стейт машину. У нее в каждый конкретный момент есть стейт, ничего искать не надо. Стейт умеет обрабатывать некоторый набор пакетов. Механизм сессий и коннекшнов тоже писать не стоит. В веб-сокетах это есть и заработает из коробки

KCherkasov committed:
implemented requested changes

-class fields placed in class' beginnings
-changed LinkedList to Deque in declaration
-refactored logging in AbstractState derivatives

Badya commented on src/main/java/statemachine/StateStack.java:
```diff
@@ -6,10 +6,14 @@
 
 public class StateStack implements PendingStack {
+    private final Deque<PendingChange> pendingChanges = new LinkedList<>();
+    private final Stack<State> stack = new Stack<>();
```
`Stack` это по сути уже тоже атавизм - использовать также `Deque` 
JavaDoc:
> A more complete and consistent set of LIFO stack operations is
>  provided by the {@link Deque} interface and its implementations, which
> should be used in preference to this class.

В качестве реализации можно взять `ArrayDeque`

KCherkasov commented:
Так и планируется сделать (коннекшн - дурное рабочее название, обозначающее класс, объединяющий в себе стейт-машину для конкретного пользователя, его данные (чтобы каждый раз не стучаться в БД, например, активный персонаж с его снаряжением, ид подземелья или группы, к которой принадлежит персонаж (тут надо подумать, как выгоднее), т.е. вещи, которые будут регулярно меняться в течение сессии и которые выгоднее держать в памяти, периодически сохраняя изменения в БД), т.е. по сути продвинутый контроллер, который не только хранит стейт-машину, но и данные, которые нужны часто, но для предотвращения читерства не следует держать только у пользователя (т.е. закешировать, например, ответ про персонажа и его вещи на фронте для ускорения отображения - вполне ок, но валидным будет только хранимое на сервере состояние)

Механизм сессий писать никто и не планировал - идея в том, чтобы использовать ид сессии как ключ для ассоциативного массива описанных выше объектов (назовем их SmartController, т.к. по сути в них инкапсулируются конкретные контроллеры в виде стейтов, плюс функция хранения полезных данных). 
Первой мыслью было использовать как ключ id юзера, но проблема в том, что по факту всякий новый посетитель - неавторизованный юзер, т.е. ключ может быть неуникален. 
Поэтому хочется гарантированно уникальный ключ, как, например id сессии (между id сессии и физическим пользователем вполне однозначная связь, и разлогинивание без дисконнекта не убьет смартконтроллер, а лишь заставит его прочистить юзердату, стек стейт-машины и перевести ее в состояние авторизации).

Идея вытекает из вот такого (вполне вероятно, что не совсем верного) понимания работы сервера - на сервер приходят пакеты с фронта (все валится в один сокет), далее пакет идентифицируется по id сессии и бросается в соотв. смартконтроллер, который его обрабатывает при помощи своей стейт-машины (в результате вполне может поменяться состояние хранимых в смартконтроллере данных и стейт-машины), формирует ответ, который бросает обратно на фронт (все опять через один сокет), а уже фронт разбирается, какому клиенту, какой пакет адресован, и обновляет для них представления.

Т.е. идея с созданием смартконтроллеров - распараллеливание нагрузки (пока реализовать просто как объектный пул, потом можно подумать про многопоточность, например группировать по Х контроллеров в поток),

Но вопрос, насколько это согласуется с идеей REST? Или использовать стейт как контроллер - нормально?

KCherkasov committed:
implemented requested changes (changed Stack to Deque/ArrayDeque)

Badya commented on src/main/java/statemachine/StateStack.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private void applyPendingChanges() {
+        for (PendingChange change : this.pendingChanges) {
```
Зачем итератор, а потом очистка - из очереди можно вынимать используя `peek()` и `poll()` - таким образом гарантируется очередность и чистить ничего не надо

Badya commented on src/main/java/statemachine/StateStack.java:
```diff
@@ -0,0 +1,120 @@
+    public void update() {
+        for (State state : this.stack) {
+            LOGGER.info("state update call: ");
+            if (!state.update()) {
```
Что подразумевает апдейт? Почему дальше break?
По стеку итерируемся, но элементы там остаются?

KCherkasov replied:
апдейт - метод для циклически повторяемых действий (привычка с оффлайновых проектов все снабжать методом апдейт). Здесь, например, можно следить за временем с момента последнего запроса от юзера и разлогиниваться по таймауту/вешать статус AFK на персонажа, перекидывая стейт обратно в ожидание авторизации. И да, элементы остаются (извлекаются они по requestStatePop() или requestStackClear() от другого стейта, т.к. стейт может быть вполне себе долгоживущим (то же нахождение в городе или в подземелье) и логично оставить самому стейту знание про то, когда ему уничтожиться и какие стейты вызвать (по сути эта логика запускается при handlePacket(), например при успешной авторизации стейт, ответственный за нее, кладет в стек pop запрос (на себя) и push запрос на следующий стейт (список персонажей авторизованного игрока).

Про апдейт метод - я бы его пока оставил, окончательно его необходимость будет ясна, когда появится примерный геймплей (потому что стейты города и подземелья будут наиболее нагруженными по функционалу, и вполне возможно, что какая-то циклика там будет, самый простой пример - циклически обновлять состояние карты подземелья (ее сделать разделяемым ресурсом между всеми членами группы) на предмет очередности хода, т.е. дошел ли ход до игрока)

Касаемо break - если в стеке более одного стейта и нет нужды в том, чтобы обновлялись нижележащие - вводится break (опять же legacy с оффлайна, и вот ее резонно убрать, т.к. я не могу придумать стейт, который бы обновлялся сам и позволял еще что-то делать нижележащим. Да и вообще, по уму в стейт машине должно быть не более 1-2 стейтов в стеке единомоментно, а обновляться только верхний)

Badya replied:
Мы называем это "тик" =)
Хорошо - претензий нет

Badya commented on src/main/java/statemachine/StateStack.java:
```diff
@@ -0,0 +1,120 @@
+            return this.state;
+        }
+
+        private final StackActions action;
```
Поля тоже наверх

Badya commented on src/main/java/states/AbstractState.java:
```diff
@@ -0,0 +1,26 @@
+
+    @Override
+    public void requestStackPush(State.StateId stateId) {
+        this.stack.pushState(stateId);
```
Впринципе писать this на каждый чих не нужно

KCherkasov replied:
привычка с плюсов =) 
какой критерий, когда this надо, когда нет (кроме очевидного 
```java
class A {
    private SomeType field;

    public void setField(SomeType field) { this.field = field; }
}
``` 
и в этом духе)?

Badya replied:
Обычно this пишут как раз когда есть конфликт имен (как в примере выше), в остальных случаях можно опустить

KCherkasov replied:
Ясно, сейчас почищу код

KCherkasov committed:
implemented requested changes

KSolovyev commented:
> Идея вытекает из вот такого (вполне вероятно, что не совсем верного) понимания работы сервера - на сервер приходят пакеты с фронта (все валится в один сокет), далее пакет идентифицируется по id сессии и бросается в соотв. смартконтроллер, который его обрабатывает при помощи своей стейт-машины (в результате вполне может поменяться состояние хранимых в смартконтроллере данных и стейт-машины), формирует ответ, который бросает обратно на фронт (все опять через один сокет), а уже фронт разбирается, какому клиенту, какой пакет адресован, и обновляет для них представления.

> Т.е. идея с созданием смартконтроллеров - распараллеливание нагрузки (пока реализовать просто как объектный пул, потом можно подумать про многопоточность, например группировать по Х контроллеров в поток),


> Но вопрос, насколько это согласуется с идеей REST? Или использовать стейт как контроллер - нормально?

Запросы не будем параллельно обрабатывать. Вы этого не умеете и получится с первого раза плохо. Будем параллелить игровые сессии. Если сессий нет и у вас открытый игровой мир, то будем думать, как это лучше сделать. Но принцип будет тот же: запросы игроков, которые могут теоретически конкурировать, будем делать в одном потоке

KCherkasov commented:
"Настоящего" открытого мира не планируется.

Есть город, который совмещает в себе функцию лобби (поиск/формирование группы), торговли (набор вендоров, которым можно продать ненужное и что-нибудь купить) и выбора квестов (решено квесты жестко связывать с подземельями, т.е. подземелье генерится под взятое задание, хоть каждый раз и случайно, но необходимые элементы для выполнения квеста там размещаются гарантированно). 

Условно "сессией" можно назвать поход в подземелье, который завершается либо по выходу из него, либо смертью всех персонажей, либо победой над всеми монстрами. В любом из трех случаев группа перемещается обратно в город (в первых двух будет возможность повторить попытку, но подземелье будет сгенерировано по-новой).

Пока из этого ясно только то, что город будет разделяемым ресурсом для игроков вне подземелья (вот как тут снизить нагрузку и расслоение? Фазирование?), а каждое подземелье - разделяемый ресурс для группы игроков.

KCherkasov committed:
temporary dump

KCherkasov committed:
Merge branch 'refs/heads/dev-tmp' into develop

KCherkasov committed:
Added unit test for StateStack using JUnit

test covers enqueueing operations, pending changes application and
update() method call

KCherkasov committed:
missing semicolon fixup

KCherkasov committed:
some more refactoring (long lines split in multiple short ones)

KCherkasov commented:
Из текущего вроде все заимплементил кроме переделки StateFactory в спринговый сервис

Badya commented on src/main/java/statemachine/StateStack.java:
```diff
@@ -0,0 +1,284 @@
+        }
+    }
+
+    public static class StateStackTest {
```
Тесты распологать в `src/test/java/...` отдельными классами

Badya commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,20 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(Main.class);
+
+    public static void main(String[] args) {
+        final Result testResult = JUnitCore.runClasses(StateStack.StateStackTest.class);
```
Запускать тесты нужно через `mvn test`, если они расположены в правильной директории `src/test/java/...`

KCherkasov replied:
Понял, этот сделал специально внутренним классом, чтобы поконтролить еще и механизмы методов и состояние полей (выловил пару экземпляров багла пока тестил).

Сейчас переделаю в нормальный тест (придется добавить малость геттеров в StateStack для сбора информации)

KSolovyev commented:
Вы же помните, что нужно сделать API? =)
И еще просьба, поправьте соседний пулл реквест

KCherkasov committed:
StateStack unit test

unit test placed in separate file in src/tests/java

added some get methods in StateStack to make it bit more verbose for
testing

KCherkasov commented:
Да, конечно =) 

Как раз коллега и занимается работой над контроллерами, пока я со стейт машиной разбираюсь =)

Badya commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,16 @@
+public class Main {
+    private static final Logger LOGGER = LoggerFactory.getLogger(Main.class);
+
+    public static void main(String[] args) {
```
Метод мейн вам вообще не нужен как и класс - его просто быть не должно - у вас есть спринг и тесты=)

KCherkasov replied:
касаемо класса мейн - таки нужен, без него тревис не пропускает тест =)

Badya commented on src/test/java/StateStackTest.java:
```diff
@@ -0,0 +1,161 @@
+public class StateStackTest {
+    private StateStack stack;
+    private Random random = new Random();
+    private static final Logger TEST_LOGGER = LoggerFactory.getLogger(StateStackTest.class);
```
Логирование в тестах можно опустить

Badya commented on src/test/java/StateStackTest.java:
```diff
@@ -0,0 +1,161 @@
+        TEST_LOGGER.info("====");
+    }
+
+    public static void main(String[] args) {
```
Мейн не нужен если запускать через `mvn test`

Badya commented on src/test/java/StateStackTest.java:
```diff
@@ -0,0 +1,161 @@
+        assertTrue("there shall be some pending changes after adding", stack.hasPendingChanges());
+        assertEquals("there shall be " + Integer.toString(testsCount) + " pending changes",
+                testsCount, stack.pendingChangesCount() );
+        stack.clearState();
```
Вот это повторяется от теста к тесту - хотя перед каждым тестом в `@Before` вы создаете новый стек.
Можно ничего не чистить или вынести это в `@After`

KCherkasov committed:
tests refactored; Main class deleted; .gitignore upd

KCherkasov committed:
main class stub returned;

KCherkasov commented:
> Идея со стейтами - хорошая, но фабрика должна быть Спринговым сервисом, а не хелпером со статическими методами. Иначе прокидывать зависимости будет очень сложно

по поводу превращения StateFactory в спринговый сервис - не совсем понял идею, в чем плюсы, если она занимается по сути только инстанцированием разных объектов? В чем будет сложность с прокидыванием зависимостей (учитывая что внутриигровой процесс, как я думаю, будет вертеться в основном вокруг написанных нами сущностей, и подключать их мы будем не к фабрике, а к конкретному классу)?

KSolovyev commented:
ну а как вы будете сообщения пользователям отправлять?
